openfe/setup/scorer.py

```diff
@@ -1,4 +1,71 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/openfe
+
+"""
+Scorers
+=======
+
+A scorer is used to determine a score (and optionally, annotations) for a
+given AtomMapping.
+"""
+
+from typing import NamedTuple, Dict, Any, Union
+
+
+class ScoreAnnotation(NamedTuple):
+    """Container for a score from a mapping and any associated annotations.
+
+    Parameters
+    ----------
+    score : float
+        The score, or ``None`` if there is no score associated
+    annotation : Dict[str, Any]
+        Mapping of annotation label to annotation value for any annotations
+    """
+    score: Union[float, None]
+    annotation: Dict[str, Any]
+
+
 class Scorer:
-    def __call__(self, atommapping):
-        return 0.0
+    """Abstract base class for Scorers.
+
+    To implement a subclass, you must implement the ``score`` method. If
+    your ``Scorer`` only returns annotations, then return None from the
+    ``score`` method. You may optionally implement the ``annotation``
+    method. The default ``annotation`` returns an empty dictionary.
+
+    Use a ``Scorer`` by calling it as a function.
+    """
+    def score(self, atommapping) -> Union[float, None]:
+        """Calculate the score for an AtomMapping.
+
+        Parameters
+        ----------
+        atommapping : AtomMapping
+            AtomMapping to score
+
+        Returns
+        -------
+        Union[float, None] :
+            The score, or ``None`` if no score is calculated
+        """
+        raise NotImplementedError()
```

Comment:
I know this _shouldn't_ be ever used on it's own, but for the sake of someone doing something silly + coverage can we a) get a test that triggers this, b) a simple message to tell folks to not call the base class directly?

Reply:
I usually use the rule that `raise NotImplementedError` is part of the coverage exclude lines (along with `__repr__`); do we want to make a practice of always testing `NotImplementedError`s? (I'm fine either way.) Will definitely add a clearer message, though.

Reply:
I'm not super fussed either way, I do find trying to trigger errors cathartic so I'm happy just adding them randomly on a bored weekend evening.

Reply:
I'll went ahead and added the test -- it isn't hard and doesn't hurt anything. We can always ignore in coverage later (having a test that covers lines that don't count toward coverage isn't a problem).

Speaking of coverage, CodeCov is being flaky. I'm trying to dig up some old CodeCov problems I've had (and their solutions) to see if it is something I've seen before.